pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -803,7 +821,9 @@ func applyTrafficPolicy(opts buildClusterOpts, proxy *model.Proxy) {
 	applyOutlierDetection(opts.cluster, outlierDetection)
 	applyLoadBalancer(opts.cluster, loadBalancer, opts.port, proxy)
 	if opts.clusterMode != SniDnatClusterMode {
-		tls = conditionallyConvertToIstioMtls(tls, opts.serviceAccounts, opts.sni, opts.proxy, opts.serviceMTLSMode)
+		// AutoMTLS and service.MTLSReady must both be true to enable
+		mtlsReady := opts.env.Mesh.GetEnableAutoMtls().Value && opts.mtlsReady
+		tls = conditionallyConvertToIstioMtls(tls, opts.serviceAccounts, opts.sni, opts.proxy, opts.serviceMTLSMode, mtlsReady)
```

Comment:
collapse args into a single tls, opts. Make this function return two values (tls context and use_intelligent_mtls) or make this function set it into the cluster opts

Reply:
pass this opts to the applyUpstreamTLSSettings. In that function, if mode is mutual and intelligent_mtls is true, then setup the cluster with two transport contexts (mtls and plaintext). else set it up with just one.